theano/sandbox/cuda/opt.py

```diff
@@ -445,6 +446,26 @@ def local_gpu_dot_to_dot22(node):
                                                 shape_out))]
     return False
 
+@local_optimizer(None)
+def local_assert_no_cpu_op(node):
+    if not isinstance(node.op, GpuOp) and all([var.owner and isinstance(var.owner.op,
+        HostFromGpu) for var in node.inputs]) and any([[c for c in var.clients
+                if isinstance(c[0].op, GpuFromHost)] for var in node.outputs]):
+            if config.assert_no_cpu_op == "warn":
+                _logger.warning(("CPU op %s is detected in the computational"
+                                 " graph") % node)
+            elif config.assert_no_cpu_op == "raise":
+                raise RuntimeError("The op %s is on CPU." % node)
+            elif config.assert_no_cpu_op == "pdb":
+                pdb.set_trace()
+
+    return None
+
+#Register the local_assert_no_cpu_op:
+assert_no_cpu_op = theano.tensor.opt.in2out(local_assert_no_cpu_op, name='assert_no_cpu_op')
+# 48.7 is after specialize device
+theano.compile.optdb.register('assert_no_cpu_op', assert_no_cpu_op, 48.7)
```

Comment:
> In order to trigger a failure with the current optimization, you could use only one elemwise       > operation with one input, for instance tanh(ms).dot(ms.T).

Yes that works.

>Then the fusion is done (at 49), and a second round of gpu transfer happens at 49.1, but then > it is too late.

Why is the second round of GPU transfers happen at 49.1, isn't the GPU specialization should happen at 48.5?